Use the SiteDiscriminator of the StaticAsset when building resource name

regcole opened the pull request:
- Addresses BroadleafCommerce/QA#1473
- tied to https://github.com/BroadleafCommerce/BroadleafCommercePrivate/pull/318
- also tied to https://github.com/BroadleafCommerce/MultiTenant-SingleSchema/pull/137
  <h3>Setup</h3>
  Install and configure Amazon module. 
  example of `common-shared.properties` used
  <b>**NOTE***</b> AWS has a problem with java 8; use java 7 when running AWS module

```
# Amazon
aws.s3.accessKeyId=(get a key)
aws.s3.secretKey=(get your own)
aws.s3.defaultBucketName=heatclinic
aws.s3.bucketSubDirectory=img
aws.s3.defaultBucketRegion=us-west-2
aws.s3.endpointURI=http://s3.amazonaws.com
```

<h2>Steps to recreate </h2>
1. Add a new image/media asset to a bundle1 on the template site (global.blc.dev )
2. promote and approve
3. delete the image from the temp directory: (example: mine was in /Library/apache-tomcat-7.0.54/temp/)
4. Go to the child site and try to view the product

<b>expected result:</b> Image should render 

<h2>Source of problem</h2>

Amazon module builds/generates an incorrect file path. In the above example, the image was placed at `bucket/img/site--1/newImage` but the child site attempted to retrieve it from path `bucket/img/site--3/newImage`. 

<h3> Solution </h3>

The system will first check the local filesystem for the static asset. If it does not find it locally, it will query the database for the corresponding `StaticAsset` object and use that to build the resource name expected by the S3 filesystem.

Commit:
Use the SiteDiscriminator of the StaticAsset when building resource name.

- Addresses BroadleafCommerce/QA#1473

phillipuniverse commented:
This should not be using the static asset service at all. This should only be dealing with files and the `FileSystemService`. Adding in the CMS module should not be apart of the solution here.

What was the problem with compiling on Java 8?

regcole commented:
@phillipuniverse  this was the Java 8 problem: https://github.com/aws/aws-sdk-java/issues/444 and more specifically: http://stackoverflow.com/questions/32058431/aws-java-sdk-aws-authentication-requires-a-valid-date-or-x-amz-date-header . It mainly just an issue with java 8 and the version of aws sdk we are using

Commit:
Iterate through all possible resource path names

- Addresses BroadleafCommerce/QA#1473

phillipuniverse commented on src/main/java/org/broadleafcommerce/vendor/amazon/s3/S3FileServiceProvider.java:
```diff
@@ -80,15 +90,42 @@ public File getResource(String name) {
             AmazonS3Client s3 = getAmazonS3Client(s3config);
-            S3Object object = s3.getObject(new GetObjectRequest(s3config.getDefaultBucketName(), buildResourceName(name)));
-            
+            String resourceName = getResourceName(name);
```
@regcole it seems like you're just finding this value and throwing it away. Isn't what you really want the `getPossibleResourceNames()` that you have below? It seems like this doesn't do anything, along with the `StringUtils.isEmpty(resourceName)`; I can't figure out the function, since `getPossibleResourceNames()` will return the site-specific resource name along with the parent's resource name

regcole replied:
@phillipuniverse  line 100 is an attempt to resolve the resource path on the local filesystem. If the resource name is found, then we know exactly what resource name to query the S3 file server for. If the resource path can not be found on the local filesystem, then the `getResourceName()` function will return null ( indicating that no resource was found). It is only at this point that the system will gather all possible resource names and query the S3 file server with all possible until it finds a match.

phillipuniverse replied:
That doesn't make sense to me. This is how I understand what you said: "Get the resource name as it is represented on the local file system. Now that we have the resource name as it is on the local filesystem, use that resource name to look it up on the remote file system (amazon)."

Why does a local resource name have relevance to the remote resource name?

regcole replied:
I believe i see the shortcomings now. If i have the resource on the local filesystem, then there was no need to look it up on the remote server.

phillipuniverse commented:
Tests fail on this feature branch. I think that in general you have the right idea.

Commit:
Only make remote calls when resource not on local filesystem

- Make sure module passes all unit test
- correct the properties file location for unit test

- Addresses BroadleafCommerce/QA#1473
- Addresses https://github.com/BroadleafCommerce/blc-amazon/pull/7/files#r54744171

Commit:
Removed irrelevant code

- reverted changes to unit test config

- Addresses BroadleafCommerce/QA#1473
